otlp/common.go

```diff
@@ -640,7 +661,45 @@ func getSampleRate(attrs map[string]interface{}) int32 {
 		sampleRate = defaultSampleRate
 	}
 	delete(attrs, sampleRateKey) // remove attr
-	return sampleRate
+	return sampleRate, true
+}
+
+// getSampleRateFromOTelSamplingThreshold returns the sampling threshold from the OpenTelemetry
+// trace state. Sampling threshold is a string that represents the sampling
+// probability using the OpenTelemetry sampling probability format.
+//
+// See https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/tracestate-probability-sampling.md
+func getSampleRateFromOTelSamplingThreshold(traceState string) (int32, bool) {
+	// split the trace state into key-value pairs
+	kvPairs := strings.Split(traceState, ";")
+	pairs := make(map[string]string, len(kvPairs))
+	for _, kv := range kvPairs {
+		kvPair := strings.Split(kv, "=")
+		if len(kvPair) != 2 {
+			continue
+		}
+		pairs[kvPair[0]] = kvPair[1]
+	}
+	// get the tvalue from the trace state pairs
+	th, ok := pairs["th"]
+	if !ok {
+		return defaultSampleRate, false
+	}
+	// get the sampling threshold
+	t, err := sampling.TValueToThreshold(th)
+	if err != nil {
+		return defaultSampleRate, false
+	}
+
+	// randomly sample between the celling and floor values of the adjusted count
```

Comment:
Statistically, this works out to either an integer if the adjusted count is exact (powers of 2) or a half-integer if the adjusted count is not exact. 

It means that people trying to send, say, 80% of their data will see our average numbers at 75% instead, and they will think that something is wrong.

I'm about to push a friendly commit that I hope you'll agree with.

Reply:
Yeah, please feel free to contribute an alternative solution.